Multiple Linear Regression


In [ ]:
# 1.importing libraries
import numpy as nm
import matplotlib.pyplot as mtp
import pandas as pd
import numpy as np

In [ ]:
# 2.importing dataset
data_set= pd.read_csv('/content/50_CompList.csv')
print(data_set.to_string())


    R&D Spend  Administration  Marketing Spend       State     Profit
0   165349.20       136897.80        471784.10    New York  192261.83
1   162597.70       151377.59        443898.53  California  191792.06
2   153441.51       101145.55        407934.54     Florida  191050.39
3   144372.41       118671.85        383199.62    New York  182901.99
4   142107.34        91391.77        366168.42     Florida  166187.94
5   131876.90        99814.71        362861.36    New York  156991.12
6   134615.46       147198.87        127716.82  California  156122.51
7   130298.13       145530.06        323876.68     Florida  155752.60
8   120542.52       148718.95        311613.29    New York  152211.77
9   123334.88       108679.17        304981.62  California  149759.96
10  101913.08       110594.11        229160.95     Florida  146121.95
11  100671.96        91790.61        249744.55  California  144259.40
12   93863.75       127320.38        249839.44     Florida  141585.52
13   91992.39       

In [ ]:
# 3.Extracting Independent and dependent
x= data_set.iloc[:, :-1].values
y= data_set.iloc[:, 4].values
df2=pd.DataFrame(x)
print("X=")
print(df2.to_string())
df3=pd.DataFrame(y)
print("Y=")
print(df3.to_string())

X=
            0          1          2           3
0    165349.2   136897.8   471784.1    New York
1    162597.7  151377.59  443898.53  California
2   153441.51  101145.55  407934.54     Florida
3   144372.41  118671.85  383199.62    New York
4   142107.34   91391.77  366168.42     Florida
5    131876.9   99814.71  362861.36    New York
6   134615.46  147198.87  127716.82  California
7   130298.13  145530.06  323876.68     Florida
8   120542.52  148718.95  311613.29    New York
9   123334.88  108679.17  304981.62  California
10  101913.08  110594.11  229160.95     Florida
11  100671.96   91790.61  249744.55  California
12   93863.75  127320.38  249839.44     Florida
13   91992.39  135495.07  252664.93  California
14  119943.24  156547.42  256512.92     Florida
15  114523.61  122616.84  261776.23    New York
16   78013.11  121597.55  264346.06  California
17   94657.16  145077.58  282574.31    New York
18   91749.16  114175.79  294919.57     Florida
19    86419.7  153514.11        0.0  

In [ ]:
# 4.Catgorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer
labelencoder_x= LabelEncoder()
x[:, 3]= labelencoder_x.fit_transform(x[:,3])
dt=pd.DataFrame(x)
print("--------------------")
print(dt.to_string())
print("-----------------------")

--------------------
            0          1          2  3
0    165349.2   136897.8   471784.1  2
1    162597.7  151377.59  443898.53  0
2   153441.51  101145.55  407934.54  1
3   144372.41  118671.85  383199.62  2
4   142107.34   91391.77  366168.42  1
5    131876.9   99814.71  362861.36  2
6   134615.46  147198.87  127716.82  0
7   130298.13  145530.06  323876.68  1
8   120542.52  148718.95  311613.29  2
9   123334.88  108679.17  304981.62  0
10  101913.08  110594.11  229160.95  1
11  100671.96   91790.61  249744.55  0
12   93863.75  127320.38  249839.44  1
13   91992.39  135495.07  252664.93  0
14  119943.24  156547.42  256512.92  1
15  114523.61  122616.84  261776.23  2
16   78013.11  121597.55  264346.06  0
17   94657.16  145077.58  282574.31  2
18   91749.16  114175.79  294919.57  1
19    86419.7  153514.11        0.0  2
20   76253.86   113867.3  298664.47  0
21   78389.47  153773.43  299737.29  2
22   73994.56  122782.75  303319.26  1
23   67532.53  105751.03  304768.73  1
24  

In [ ]:
# 5.State column
ct = ColumnTransformer([("State", OneHotEncoder(), [3])], remainder =
'passthrough')
x = ct.fit_transform(x)
""" We should not use all the dummy variables at the same time, so it
must be 1 less than the total number of
dummy variables, else it will create a dummy variable trap."""
print("----------b4 removing dummy variable-----")
dfx=pd.DataFrame(x)
print("--------------------")
print(dfx.to_string())
print("-----------------------")

----------b4 removing dummy variable-----
--------------------
      0    1    2          3          4          5
0   0.0  0.0  1.0   165349.2   136897.8   471784.1
1   1.0  0.0  0.0   162597.7  151377.59  443898.53
2   0.0  1.0  0.0  153441.51  101145.55  407934.54
3   0.0  0.0  1.0  144372.41  118671.85  383199.62
4   0.0  1.0  0.0  142107.34   91391.77  366168.42
5   0.0  0.0  1.0   131876.9   99814.71  362861.36
6   1.0  0.0  0.0  134615.46  147198.87  127716.82
7   0.0  1.0  0.0  130298.13  145530.06  323876.68
8   0.0  0.0  1.0  120542.52  148718.95  311613.29
9   1.0  0.0  0.0  123334.88  108679.17  304981.62
10  0.0  1.0  0.0  101913.08  110594.11  229160.95
11  1.0  0.0  0.0  100671.96   91790.61  249744.55
12  0.0  1.0  0.0   93863.75  127320.38  249839.44
13  1.0  0.0  0.0   91992.39  135495.07  252664.93
14  0.0  1.0  0.0  119943.24  156547.42  256512.92
15  0.0  0.0  1.0  114523.61  122616.84  261776.23
16  1.0  0.0  0.0   78013.11  121597.55  264346.06
17  0.0  0.0  1.0  

In [ ]:
#6.avoiding the dummy variable trap:
x = x[:, 1:]
df4=pd.DataFrame(x)
print("Updated X=")
print(df4.to_string())

Updated X=
      0    1          2          3          4
0   0.0  1.0   165349.2   136897.8   471784.1
1   0.0  0.0   162597.7  151377.59  443898.53
2   1.0  0.0  153441.51  101145.55  407934.54
3   0.0  1.0  144372.41  118671.85  383199.62
4   1.0  0.0  142107.34   91391.77  366168.42
5   0.0  1.0   131876.9   99814.71  362861.36
6   0.0  0.0  134615.46  147198.87  127716.82
7   1.0  0.0  130298.13  145530.06  323876.68
8   0.0  1.0  120542.52  148718.95  311613.29
9   0.0  0.0  123334.88  108679.17  304981.62
10  1.0  0.0  101913.08  110594.11  229160.95
11  0.0  0.0  100671.96   91790.61  249744.55
12  1.0  0.0   93863.75  127320.38  249839.44
13  0.0  0.0   91992.39  135495.07  252664.93
14  1.0  0.0  119943.24  156547.42  256512.92
15  0.0  1.0  114523.61  122616.84  261776.23
16  0.0  0.0   78013.11  121597.55  264346.06
17  0.0  1.0   94657.16  145077.58  282574.31
18  1.0  0.0   91749.16  114175.79  294919.57
19  0.0  1.0    86419.7  153514.11        0.0
20  0.0  0.0   76253.86

In [ ]:
# 7.Splitting the dataset into training and test set
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test= train_test_split(x, y, test_size= 0.2, random_state=0)

In [ ]:
# 8.Fitting the MLR model to the training set:
from sklearn.linear_model import LinearRegression
regressor= LinearRegression()
regressor.fit(x_train, y_train)

LinearRegression()

In [ ]:
# 9.Predicting the Test set result;
y_pred=regressor.predict(x_test)

In [ ]:
# 10.To compare the actual output values for X_test with the predicted value
df = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})
print(df.to_string())
print("Mean")
print(data_set.describe())
print("-------------------------------------")

      Actual      Predicted
0  103282.38  103015.201598
1  144259.40  132582.277608
2  146121.95  132447.738452
3   77798.83   71976.098513
4  191050.39  178537.482211
5  105008.31  116161.242302
6   81229.06   67851.692097
7   97483.56   98791.733747
8  110352.25  113969.435330
9  166187.94  167921.065696
Mean
           R&D Spend  Administration  Marketing Spend         Profit
count      50.000000       50.000000        50.000000      50.000000
mean    73721.615600   121344.639600    211025.097800  112012.639200
std     45902.256482    28017.802755    122290.310726   40306.180338
min         0.000000    51283.140000         0.000000   14681.400000
25%     39936.370000   103730.875000    129300.132500   90138.902500
50%     73051.080000   122699.795000    212716.240000  107978.190000
75%    101602.800000   144842.180000    299469.085000  139765.977500
max    165349.200000   182645.560000    471784.100000  192261.830000
-------------------------------------


In [16]:
# 11.Evaluating the Algorithm
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 7514.293659636814
Mean Squared Error: 83502864.03250548
Root Mean Squared Error: 9137.990152791011


You can see that the value of root mean squared error is 9137.99,
which is less than 10% of the mean value
of the expenses in all states. This means that our algorithm is very
accurate and good predictions.

In [18]:
# 12.predicting the accuracy score
from sklearn.metrics import r2_score
score=r2_score(y_test,y_pred)
print("r2 socre is ",score*100,"%")

r2 socre is  93.47068473282987 %
